airflow/models/xcom_arg.py

```diff
@@ -332,7 +341,23 @@ def get_task_map_length(self, run_id: str, *, session: Session) -> int | None:
 
     @provide_session
     def resolve(self, context: Context, session: Session = NEW_SESSION) -> Any:
-        ti = context["ti"]
+        from airflow.models.taskinstance import TaskInstance
+
+        context_ti = context["ti"]
+        if isinstance(context_ti, TaskInstance):
+            ti = context_ti
+        else:
+            ti = (
+                session.query(TaskInstance)
+                .filter_by(
+                    dag_id=context_ti.dag_id,
+                    task_id=context_ti.task_id,
+                    run_id=context_ti.run_id,
+                    map_index=context_ti.map_index,
+                )
+                .one()
+            )
+
```

Comment:
This part is entirely unrelated, but Mypy is blocking me since` context["ti"]` can be a TaskInstancePydantic.

Reply:
Prepend  it with `assert(isinstance(ti, TaskInstance)` this will satisfy MyPy and will allow us to implement AIP-44 veriant when our tests will start failing with the assert. (and assert will be optimized away).

Reply:
(For us it is important as this will also flag all the cases where we will have to implement AIP-44 equivalent, so having assert here is really helpful, and I am glad it actually is needed here- that was exactly the intention of adding 
 TaskInstancePydantic now to Context)